subsys/bluetooth/audio/has.c

```diff
@@ -34,11 +34,86 @@ static ssize_t read_features(struct bt_conn *conn, const struct bt_gatt_attr *at
 				 sizeof(has.features));
 }
 
+#if CONFIG_BT_HAS_PRESET_COUNT > 0
+static uint8_t handle_control_point_op(struct bt_conn *conn, struct net_buf_simple *buf)
+{
+	const struct bt_has_cp_hdr *hdr;
+
+	hdr = net_buf_simple_pull_mem(buf, sizeof(*hdr));
+
+	BT_DBG("conn %p opcode %s (0x%02x)", conn, bt_has_op_str(hdr->opcode), hdr->opcode);
+
+	/* TODO: handle request here */
+
+	return BT_HAS_ERR_INVALID_OPCODE;
+}
+
+static ssize_t write_control_point(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				   const void *data, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	struct net_buf_simple buf;
+	uint8_t err;
+
+	BT_DBG("conn %p attr %p data %p len %d offset %d flags 0x%02x", conn, attr, data, len,
+	       offset, flags);
+
+	if (offset > 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (len == 0) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	net_buf_simple_init_with_data(&buf, (void *)data, len);
+
+	err = handle_control_point_op(conn, &buf);
+	if (err) {
+		BT_WARN("err 0x%02x", err);
+		return BT_GATT_ERR(err);
+	}
+
+	return len;
+}
+
+static ssize_t read_active_preset_index(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+					void *buf, uint16_t len, uint16_t offset)
+{
+	BT_DBG("conn %p attr %p offset %d", conn, attr, offset);
+
+	if (offset > sizeof(has.active_index)) {
```

Comment:
">=" ?
If the size is 2, and the offset is 2, the read will be outside of the value.  (For a size of 2, offsets 0 and 1 are OK.)

Reply:
On the other hand, I don't think HAS supports the read long procedure at all, does it? In that case, we should not allow offset to be anything but 0.

Reply:
@MariuszSkamra Does the spec do zero-based indexing?
I think the intention here is exactly what I am saying - if the offset is outside of the size of the value to be read, return an error.
(Alternatively, it may be that they allow "one past" - will have to check with someone.)

Reply:
> (Alternatively, it may be that they allow "one past" - will have to check with someone.)

If you have a value of size 2 (16-bit) and someone does a read on offset 2, then you can just return 0 bytes. I don't think that is an issue.

Reply:
@asbjornsabo There's qualification test for it:
GATT/SR/GAR/BV-08-C [Read Behind Long Characteristic Descriptor - from Server]
"Verify that a Generic Attribute Profile server returns a 0-length part value when reading a Long Characteristic Descriptor selected by handle with an offset equal to the length of the Long Characteristic Value."

Reply:
Yes.  I checked with Pål yesterday, and there is a requirement that "If the value offset of the ATT_READ_BLOB_REQ PDU is equal to the length of the attribute value, then the length of the part attribute value in the response shall be zero."
So, the code is correct and I have learnt something :-)

Reply:
(Feel free to resolve - I leave it open to make it easy to see my reply.)